### PR TITLE
Allow BindPropertyAttribute to be specified on controller classes

### LINKED ISSUE #7350: `[BindProperty]` does not work on controller classes
The [ASP.NET Core 2.1 Roadmap](https://blogs.msdn.microsoft.com/webdev/2018/02/02/asp-net-core-2-1-roadmap/#mvc) mentions placing `[BindProperty]` on a controller class will bind all properties in the controller. This is not the case.

For example, with
``` c#
public class VController : Controller
{
    [BindProperty(SupportsGet = false)]
    public string FirstName { get; set; } = "John";

    [BindProperty(SupportsGet = false)]
    public string LastName { get; set; } = "Smith";

    // ...
}
```
both `FirstName` and `LastName` are model bound. With
``` c#
[BindProperty(SupportsGet = false)]
public class VController : Controller
{
    public string FirstName { get; set; } = "John";

    public string LastName { get; set; } = "Smith";

    // ...
}
```
neither property is bound.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #7350

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/BindPropertyAttributeApplicationModelProvider.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Mvc.ApplicationModels;
+using Microsoft.AspNetCore.Mvc.Internal;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace Microsoft.AspNetCore.Mvc
+{
+    internal class BindPropertyAttributeApplicationModelProvider : IApplicationModelProvider
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/BindPropertyAttributeApplicationModelProvider.cs]: Generally we'd try to do this in the default provider. Is there a really good reason why this is separate?

Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/BindPropertyAttributeApplicationModelProvider.cs]: I like the idea of discovery happening in stages, but that's the not the existing design. I don't want to do 10% of it differently.

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/BindPropertyAttributeApplicationModelProvider.cs]: Not a particularly good reason. I just wanted to treat it like some of the other property attribute providers we have like `TempData` \ `ViewData`.

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/BindPropertyAttributeApplicationModelProvider.cs]: That's how it's done for pages too. That said, we basically allow any binding specific metadata on Razor Page handler to apply to the property:

https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageApplicationModelProvider.cs#L268-L276

IMO, specifically looking for `BindPropertyAttribute` seems like a safer \ better thing to do than indiscriminately pick metadata from the controller.

**Comment by pranavkm** [comment]
🆙 📅

**Comment by pranavkm** [comment]
🆙 📅

**Comment by pranavkm** [comment]
Bump

**Pull request merged**